### PR TITLE
[New RM] Add DuckDuckGoRM

### PR DESCRIPTION
**Pull request by zenith110**
This PR aims to accomplish the following:

- Creates a retriever model using DuckDuckGo api utilizing the [duckduckgo_search](https://github.com/deedy5/duckduckgo_search) library.
- Update all the LM examples to include DuckDuckGo as a retriever in examples/
- Added documentation for new RM, as well as custom options to enable safe search/language.

Input for this example was pokemon.

Output results:
```python
python examples/run_storm_wiki_gpt.py --output-dir ./output/ddg --retriever duckduckgo --do-research --do-generate-outline --do-generate-article --do-polish-article
primp : INFO     : response: https://duckduckgo.com/ 200 17851
primp : INFO     : response: https://duckduckgo.com/ 200 17826
primp : INFO     : response: https://duckduckgo.com/ 200 18008
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pok%C3%A9mon+popularity+worldwide&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-129249434834354548125976907575122126238&bing_market=wt-WT&ex=-1 200 28905
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=History+of+Pokemon+franchise&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-215688423644065080573760632631242516542&bing_market=wt-WT&ex=-1 200 34438
primp : INFO     : response: https://duckduckgo.com/ 200 18000
primp : INFO     : response: https://duckduckgo.com/ 200 17891
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pokemon+breeding+mechanics+in+generation+3&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-158056105470667184619806054620694982411&bing_market=wt-WT&ex=-1 200 46564
primp : INFO     : response: https://duckduckgo.com/ 200 18008
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pok%C3%A9mon+influence+in+different+cultures&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-190062255219678873094355623798618716733&bing_market=wt-WT&ex=-1 200 37121
primp : INFO     : response: https://duckduckgo.com/ 200 17967
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Key+elements+of+Pokemon+franchise&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-19184000702775355516285330326832236310&bing_market=wt-WT&ex=-1 200 35565
primp : INFO     : response: https://duckduckgo.com/ 200 17957
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pokemon+breeding+mechanics+in+generation+1&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-191662332061853487589854932556394924935&bing_market=wt-WT&ex=-1 200 47366
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pok%C3%A9mon+cultural+impact+by+country&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-332933843419043581125777214343629805443&bing_market=wt-WT&ex=-1 200 26442
primp : INFO     : response: https://duckduckgo.com/ 200 18004
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=General+overview+of+Pokemon+franchise&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-43051092562758489970523555587206032225&bing_market=wt-WT&ex=-1 200 31259
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pokemon+breeding+mechanics+in+generation+2&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-98597388395079432292487389899060676967&bing_market=wt-WT&ex=-1 200 46458
primp : INFO     : response: https://duckduckgo.com/ 200 18240
primp : INFO     : response: https://duckduckgo.com/ 200 17996
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pok%C3%A9mon+cultural+significance+in+different+countries&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-302455654261892553942620478108956667310&bing_market=wt-WT&ex=-1 200 24591
primp : INFO     : response: https://duckduckgo.com/ 200 18550
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Evolution+of+Pokemon+in+popular+culture&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-231683652380273937004558031336891212322&bing_market=wt-WT&ex=-1 200 29333
primp : INFO     : response: https://duckduckgo.com/ 200 17946
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Impact+of+international+expansion+of+Pok%C3%A9mon+on+cultural+significance&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-11200684812693598321046570499075236720&bing_market=wt-WT&ex=-1 200 23247
primp : INFO     : response: https://duckduckgo.com/ 200 18228
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pokemon+franchise+influence+on+society&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-92532925074780844166694809854872994820&bing_market=wt-WT&ex=-1 200 24826
primp : INFO     : response: https://duckduckgo.com/ 200 18150
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Introduction+of+new+Pok%C3%A9mon+in+later+generations&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-213651897599102835988391789128957127802&bing_market=wt-WT&ex=-1 200 27333
primp : INFO     : response: https://duckduckgo.com/ 200 18463
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Impact+of+Pokemon+franchise+on+popular+culture&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-184982165892032937003043399095966858452&bing_market=wt-WT&ex=-1 200 25008
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Improvements+in+breeding+mechanics+of+Pok%C3%A9mon+in+Generation+6&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-263900224743465678794443065316720644902&bing_market=wt-WT&ex=-1 200 35487
primp : INFO     : response: https://duckduckgo.com/ 200 18554
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Changes+in+breeding+mechanics+of+Pok%C3%A9mon+from+Generation+6+onwards&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-13009221313963447274061041979850108320&bing_market=wt-WT&ex=-1 200 35167
primp : INFO     : response: https://duckduckgo.com/ 200 18508
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Comparison+of+breeding+mechanics+in+different+Pok%C3%A9mon+generations&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-45837180021057475472832486922139535494&bing_market=wt-WT&ex=-1 200 36649
primp : INFO     : response: https://duckduckgo.com/ 200 18198
primp : INFO     : response: https://duckduckgo.com/ 200 18808
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pokemon+characters+contribution+to+franchise+success&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-35387584357503951149786457497573218182&bing_market=wt-WT&ex=-1 200 21981
primp : INFO     : response: https://duckduckgo.com/ 200 18169
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Impact+of+Generation+3+Pok%C3%A9mon+on+cultural+significance+in+different+countries&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-277301478597205438538401168802403074254&bing_market=wt-WT&ex=-1 200 23680
primp : INFO     : response: https://duckduckgo.com/ 200 18288
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Popular+Pokemon+characters+in+Pokemon+franchise&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-274034312682286756031617244841971097318&bing_market=wt-WT&ex=-1 200 43394
primp : INFO     : response: https://duckduckgo.com/ 200 18030
primp : INFO     : response: https://duckduckgo.com/ 200 18169
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Cultural+significance+of+Pok%C3%A9mon+in+different+countries&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-311683773483935830606258027745036181806&bing_market=wt-WT&ex=-1 200 37318
primp : INFO     : response: https://duckduckgo.com/ 200 18770
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Most+iconic+and+popular+Pokemon+characters&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-278325000143127116537133211417368449504&bing_market=wt-WT&ex=-1 200 43819
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pokemon+breeding+mechanics+evolution+Generation+6&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-102296155810643392695964731955234839610&bing_market=wt-WT&ex=-1 200 47028
primp : INFO     : response: https://duckduckgo.com/ 200 18316
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Impact+of+Generation+5+Pok%C3%A9mon+on+cultural+significance+in+different+countries&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-247879090801879403307551725275674782038&bing_market=wt-WT&ex=-1 200 23474
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Breeding+changes+in+Pokemon+games+Generation+6+and+beyond&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-302292130751198998999655957641793643635&bing_market=wt-WT&ex=-1 200 37990
primp : INFO     : response: https://duckduckgo.com/ 200 17922
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Masuda+Method+shiny+Pokemon+breeding&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-76255668511939590336435179941469566429&bing_market=wt-WT&ex=-1 200 46961
primp : INFO     : response: https://duckduckgo.com/ 200 17919
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pok%C3%A9mon+evolution+over+the+years&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-73984785480059043361813761279045821320&bing_market=wt-WT&ex=-1 200 39844
primp : INFO     : response: https://duckduckgo.com/ 200 17865
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=history+of+Pok%C3%A9mon+franchise&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-182715789477438060616785546973194145371&bing_market=wt-WT&ex=-1 200 37995
primp : INFO     : response: https://duckduckgo.com/ 200 17865
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=origin+of+Pok%C3%A9mon+characters&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-288851543332540401576252132597494345741&bing_market=wt-WT&ex=-1 200 35319
primp : INFO     : response: https://duckduckgo.com/ 200 18222
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Influence+of+Pok%C3%A9mon+franchise+on+gaming+industry&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-224649821550548649350692816370944703899&bing_market=wt-WT&ex=-1 200 23371
primp : INFO     : response: https://duckduckgo.com/ 200 18189
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Impact+of+Pok%C3%A9mon+franchise+on+popular+culture&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-146067067955270654279750736344609611392&bing_market=wt-WT&ex=-1 200 26902
primp : INFO     : response: https://duckduckgo.com/ 200 18335
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=Pok%C3%A9mon+franchise+impact+on+media+and+merchandise+industry&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-251890639533309742074905876088086220060&bing_market=wt-WT&ex=-1 200 23751
primp : INFO     : response: https://duckduckgo.com/ 200 18203
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=What+events+does+the+Pok%C3%A9mon+franchise+host%3F&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-287170741861637453879581290863906591271&bing_market=wt-WT&ex=-1 200 29876
primp : INFO     : response: https://duckduckgo.com/ 200 18513
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=How+has+the+Pok%C3%A9mon+franchise+diversified+its+revenue+streams%3F&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-308845482102921344889326992059422102475&bing_market=wt-WT&ex=-1 200 25110
primp : INFO     : response: https://duckduckgo.com/ 200 18348
primp : INFO     : response: https://links.duckduckgo.com/d.js?q=What+type+of+merchandise+does+the+Pok%C3%A9mon+franchise+sell%3F&kl=wt-wt&l=wt-wt&p=&s=0&df=&vqd=4-29594911902206125795502181166788493555&bing_market=wt-WT&ex=-1 200 31176
knowledge_storm.interface : INFO     : run_knowledge_curation_module executed in 53.6151 seconds
knowledge_storm.interface : INFO     : run_outline_generation_module executed in 7.4055 seconds
sentence_transformers.SentenceTransformer : INFO     : Use pytorch device_name: cpu
sentence_transformers.SentenceTransformer : INFO     : Load pretrained SentenceTransformer: paraphrase-MiniLM-L6-v2
.venv\Lib\site-packages\transformers\tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
knowledge_storm.interface : INFO     : run_article_generation_module executed in 20.0359 seconds
knowledge_storm.interface : INFO     : run_article_polishing_module executed in 8.5975 seconds
***** Execution time *****
run_knowledge_curation_module: 53.6151 seconds
run_outline_generation_module: 7.4055 seconds
run_article_generation_module: 20.0359 seconds
run_article_polishing_module: 8.5975 seconds
***** Token usage of language models: *****
run_knowledge_curation_module
    gpt-3.5-turbo: {'prompt_tokens': 17782, 'completion_tokens': 4599}
    gpt-4o: {'prompt_tokens': 0, 'completion_tokens': 0}
run_outline_generation_module
    gpt-3.5-turbo: {'prompt_tokens': 0, 'completion_tokens': 0}
    gpt-4o: {'prompt_tokens': 3645, 'completion_tokens': 530}
run_article_generation_module
    gpt-3.5-turbo: {'prompt_tokens': 0, 'completion_tokens': 0}
    gpt-4o: {'prompt_tokens': 6835, 'completion_tokens': 2515}
run_article_polishing_module
    gpt-3.5-turbo: {'prompt_tokens': 0, 'completion_tokens': 0}
    gpt-4o: {'prompt_tokens': 2643, 'completion_tokens': 700}
***** Number of queries of retrieval models: *****
run_knowledge_curation_module: {'DuckDuckGoRM': 36}
run_outline_generation_module: {'DuckDuckGoRM': 0}
run_article_generation_module: {'DuckDuckGoRM': 0}
run_article_polishing_module: {'DuckDuckGoRM': 0}
```

Thank you.

### CODE REVIEW

File: knowledge_storm/rm.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import logging
 import os
 from typing import Callable, Union, List
-
+from duckduckgo_search import DDGS
```
Comment (shaoyijia) [inline knowledge_storm/rm.py:4]: Could we move this to `DuckDuckGoSearchRM.__init__()`? In this way, people don't necessarily need to install this package and we don't need to change `requirements.txt`.  You can do something similar to [here](https://github.com/stanford-oval/storm/blob/53b7f1ece52c1315694bb16e689e69a654332768/knowledge_storm/lm.py#L247)

File: knowledge_storm/rm.py [hunk 2]
```diff
@@ -509,3 +509,118 @@ def forward(self, query_or_queries: Union[str, List[str]], exclude_urls: List[st
...
+        # If not None, is_valid_source shall be a function that takes a URL and returns a boolean.
+        if is_valid_source:
+            self.is_valid_source = is_valid_source
+        else:
+            self.is_valid_source = lambda x: True
+
+        # Import the duckduckgo search library found here: https://github.com/deedy5/duckduckgo_search
+        self.ddgs = DDGS()
+
+    # Store usage of credits and reset the class usage variable
```
Comment (shaoyijia) [inline knowledge_storm/rm.py:563]: Delete this line.

File: knowledge_storm/rm.py [hunk 3]
```diff
@@ -509,3 +509,118 @@ def forward(self, query_or_queries: Union[str, List[str]], exclude_urls: List[st
                 logging.error(f'Error occurs when searching query {query}: {e}')
 
         return collected_results
+
```
Comment (shaoyijia) [inline knowledge_storm/rm.py:512]: Add one more line break here.

File: knowledge_storm/rm.py [hunk 4]
```diff
@@ -509,3 +509,118 @@ def forward(self, query_or_queries: Union[str, List[str]], exclude_urls: List[st
...
+        # Import the duckduckgo search library found here: https://github.com/deedy5/duckduckgo_search
+        self.ddgs = DDGS()
+
+    # Store usage of credits and reset the class usage variable
+    def get_usage_and_reset(self):
+        usage = self.usage
+        self.usage = 0
+        return {'DuckDuckGoRM': usage}
+
+    # this is the important function that is used by Retriever in the pipeline
```
Comment (shaoyijia) [inline knowledge_storm/rm.py:569]: Delete this line.

**Review by shaoyijia — CHANGES_REQUESTED** [review]
Thanks for providing the log! It's pretty helpful. Just a few nits and I am happy to merge this.

**Comment by Yucheng-Jiang** [comment]
Thanks for resolving the comments! Just one additional nit, otherwise looks good to me.

Would you please add DuckDuckGo in [readme API section here](https://github.com/stanford-oval/storm?tab=readme-ov-file#api)?

Screenshot 2024-08-20 at 7 55 16 AM

**Review by shaoyijia — DISMISSED** [review]
Thanks!

**Review by shaoyijia — APPROVED** [review]

**Pull request merged**